File: Python/ByTier/F/1954_F_Unique_Strings.py
# Problem: CF 1954 F - Unique Strings
# https://codeforces.com/contest/1954/problem/F

"""
Algorithm: Computes the number of unique binary strings of length n with c 1s at the start
           and at most k 0s replaced with 1s, considering cyclic equivalence.

Approach: 
- Uses Burnside's lemma combined with inclusion-exclusion principle.
- Applies Euler's totient function to count distinct necklaces under rotation.
- For each divisor d of n, computes valid configurations of 1s and 0s that are
  distinct under cyclic shifts.
- Utilizes modular arithmetic and precomputed factorials and inverses for combinatorics.

Time Complexity: O(n * sqrt(n) + n * log(mod))
Space Complexity: O(n)

Techniques:
- Sieve of Euler for computing phi function
- Modular inverse using extended Euclidean algorithm
- Precomputed factorials and inverse factorials for combinations
- Inclusion-exclusion principle for counting valid rotations
"""

def solve(n, c, k, mod=10**9 + 7):
    """
    Compute the answer for given n, c, k according to the modified algorithm.
    Returns the result as an integer.
    """
    inv2 = (mod + 1) // 2
    f = [False] * (n + 1)
    phi = [0] * (n + 1)
    phi[1] = 1
    primes = []
    for i in range(2, n + 1):
        if not f[i]:
            phi[i] = i - 1
            primes.append(i)
        for p in primes:
            if i * p > n:
                break
            f[i * p] = True
            if i % p == 0:
                phi[i * p] = phi[i] * p
                break
            phi[i * p] = phi[i] * (p - 1)
    inv = [0] * (n + 1)
    inv[1] = 1
    for i in range(2, n + 1):
        inv[i] = mod - (mod // i) * inv[mod % i] % mod
    fac = [1] * (n + 1)
    ifac = [1] * (n + 1)
    for i in range(1, n + 1):
        fac[i] = fac[i - 1] * i % mod
        ifac[i] = ifac[i - 1] * inv[i] % mod

    def C(n_, k_):
        if n_ < k_ or k_ < 0:
            return 0
        return fac[n_] * ifac[k_] % mod * ifac[n_ - k_] % mod

    ans = 0
    for i in range(1, n + 1):
        if n % i:
            continue
        l, r = i - 1, i - 2
        sum_ = 0
        val = 1 if (c + k == n) else 0
        lim = min((c + k) // (n // i), i - 1)
        x, a = 1, i - c
        while x <= a and x <= i - c:
            for p in range(a + c, a - 1, -1):
                sum_ = (sum_ - C(p - x, l - 1) + C(p - x, r)) * inv2 % mod
            L = max(0, i - x - lim)
            R = i - x - c
            while l != L:
                l -= 1
                sum_ = (sum_ + C(a - x, l)) % mod
            while r != R:
                sum_ = (sum_ - C(a - x, r)) % mod
                r -= 1
            sign = 1 if (x & 1) else -1
            term = (C(a, x) + C(a - 1, x - 1) * c) % mod
            val = (val + sign * term * sum_) % mod
            x += 1
            a -= c
        ans = (ans + val * phi[n // i]) % mod
    ans = ans * ifac[n] % mod * fac[n - 1] % mod
    return (ans + mod) % mod


if __name__ == "__main__":
    import sys

    data = sys.stdin.readline().split()
    if len(data) >= 3:
        n, c, k = map(int, data[:3])
        print(solve(n, c, k))
    else:
        n, c, k = map(int, input().split())
        print(solve(n, c, k))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1954/1954_F_Unique_Strings.cpp
// Problem: CF 1954 F - Unique Strings
// https://codeforces.com/contest/1954/problem/F

/*
Code Purpose:
This code solves the problem of counting unique binary strings that can be formed from a given string of 1s and 0s, 
by replacing at most k 0s with 1s. The strings are considered equal if one can be obtained by cyclically shifting the other.
The solution uses Burnside's lemma and combinatorics to count distinct necklaces under rotation, with additional constraints on the number of operations.

Algorithms/Techniques:
- Burnside's Lemma: For counting distinct necklaces under rotation
- Combinatorics: Used for computing combinations C(n, k)
- Modular Exponentiation: For modular inverse calculations
- Memoization: To optimize recursive function computations

Time Complexity: O(n^2 * log(M)) where n is the length of the string and M is the modulo value.
Space Complexity: O(n) for storing precomputed factorials, inverse factorials, and memoization table.

*/
#include <algorithm>
#include <iostream>
#include <map>
#include <numeric>

using namespace std;
typedef long long ll;
#define M 1000000007
#define N 3000
int i, j, k, n, m, t, d;
ll jc[N + 50], inv[N + 50], p[N + 50], res, g[N + 50];
ll su(ll a, ll b) {
  a += b;
  return (a >= M) ? a - M : a;
}
ll ksm(ll a, ll p) {
  ll res = 1;
  while (p) {
    if (p & 1) {
      res = res * a % M;
    }
    a = a * a % M;
    p >>= 1;
  }
  return res;
}
ll c(ll n, ll m) {
  if (n < m || n < 0 || m < 0) {
    return 0;
  }
  return jc[n] * inv[m] % M * inv[n - m] % M;
}
map<ll, ll> f;
ll fuck(ll k, ll m, ll sz) {
  if (f.count(k))
    return f[k];
  ll res = 0, i, j;
  sz = min(sz, k);
  if (k == sz)
    res++;
  for (i = 1; i * m <= sz; i++) {
    ll t1 = (c(k - i * m, i) + m * c(k - i * m - 1, i - 1)) % M, t2 = 0;
    for (j = 0; j <= sz - i * m; j++)
      t2 += c(k - i * (m + 1), j);
    t1 = t1 * (t2 % M) % M;
    if (i & 1)
      res += t1;
    else
      res -= t1;
  }
  res = (res % M + M) % M;
  return f[k] = res;
}
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  jc[0] = inv[0] = 1;
  for (i = 1; i <= N; i++)
    jc[i] = jc[i - 1] * i % M;
  inv[N] = ksm(jc[N], M - 2);
  for (i = N - 1; i >= 1; i--)
    inv[i] = inv[i + 1] * (i + 1) % M;
  cin >> n >> m >> t;
  for (i = 1; i <= n; i++) {
    k = gcd(i, n);
    res += fuck(k, m, (m + t) / (n / k));
    res %= M;
  }
  cout << res * ksm(n, M - 2) % M;
}


// https://github.com/VaHiX/CodeForces/